### PR TITLE
climb frame tree to determine top level eTLD+1 when generating farble seed

### PR DESCRIPTION
**Pull request by pes10k**
This PR tries to solve a problem where we were using garbage memory sometimes when generating the farble seed, by trying to climb the frame tree. I hope i got it right…

Should fix https://github.com/brave/brave-browser/issues/10260

### CODE REVIEW

File: chromium_src/third_party/blink/renderer/core/dom/document.cc [hunk 1]
```diff
@@ -22,6 +23,31 @@
...
+  }
+
+  const auto* top_sec_ctx = top_loc_frame->Tree().Top().GetSecurityContext();
+  if (top_sec_ctx == nullptr) {
+    LOG(ERROR) << "Unable to farble frame bc can't find true top frame: "
+        << doc.Url();
+    return "";
+  }
+
+  return top_sec_ctx->GetSecurityOrigin()->RegistrableDomain().Utf8();
```
Comment (pilgrim-brave) [inline chromium_src/third_party/blink/renderer/core/dom/document.cc:48]: `RegistrableDomain` is not what we want. See examples in https://url.spec.whatwg.org/#host-registrable-domain

File: chromium_src/third_party/blink/renderer/core/dom/document.cc [hunk 2]
```diff
@@ -22,6 +23,33 @@
...
 namespace {
 
+//  Returns the eTLD+1 for the top level frame the document is in.
+//
+//  Returns the eTLD+1 (effective registrable domain) for the top level
+//  frame that the given document is in. This includes frames that
+//  are disconnect, remote or local to the top level frame.
+//
+//  If the top level frame is not discoverable, an empty string is returned.
+std::string topETLDPlusOneForDoc(const Document& doc) {
```
Comment (bridiver) [inline chromium_src/third_party/blink/renderer/core/dom/document.cc:33]: this method name should be capitalized

File: chromium_src/third_party/blink/renderer/core/dom/document.cc [hunk 3]
```diff
@@ -22,6 +23,33 @@
...
+
+  const auto* top_sec_ctx = top_loc_frame->Tree().Top().GetSecurityContext();
+  if (top_sec_ctx == nullptr) {
+    LOG(ERROR) << "Unable to farble frame bc can't find true top frame: "
+        << doc.Url();
+    return "";
+  }
+
+  const std::string host = top_sec_ctx->GetSecurityOrigin()->Host().Utf8();
+  return net::registry_controlled_domains::GetDomainAndRegistry(
```
Comment (bridiver) [inline chromium_src/third_party/blink/renderer/core/dom/document.cc:49]: we should be using `network_utils::GetDomainAndRegistry` in blink

File: chromium_src/third_party/blink/renderer/core/dom/document.cc [hunk 4]
```diff
@@ -22,6 +23,33 @@
 
 namespace {
 
+//  Returns the eTLD+1 for the top level frame the document is in.
+//
+//  Returns the eTLD+1 (effective registrable domain) for the top level
+//  frame that the given document is in. This includes frames that
+//  are disconnect, remote or local to the top level frame.
+//
+//  If the top level frame is not discoverable, an empty string is returned.
```
Comment (bridiver) [inline chromium_src/third_party/blink/renderer/core/dom/document.cc:32]: when does this happen?

File: chromium_src/third_party/blink/renderer/core/dom/document.cc (hunk 3 above)
Comment (pes10k) [reply chromium_src/third_party/blink/renderer/core/dom/document.cc]: okie, moved to network_utils::GetDomainAndRegistry

File: chromium_src/third_party/blink/renderer/core/dom/document.cc (hunk 4 above)
Comment (pes10k) [reply chromium_src/third_party/blink/renderer/core/dom/document.cc:32]: We don't know all the cases, but I updated the comment to include an example

File: chromium_src/third_party/blink/renderer/core/dom/document.cc (hunk 1 above)
Comment (pes10k) [reply chromium_src/third_party/blink/renderer/core/dom/document.cc]: fixed!

File: chromium_src/third_party/blink/renderer/core/dom/document.cc (hunk 2 above)
Comment (pes10k) [reply chromium_src/third_party/blink/renderer/core/dom/document.cc]: Done

File: chromium_src/third_party/blink/renderer/core/dom/document.cc [hunk 5]
```diff
@@ -18,10 +18,42 @@
...
+//  This can happen in cases including frames that are not connected
+//  to the document, or iframes that are accessed while they're moving
+//  between processes (e.g., a frame's url is being changed to
+//  be LocalFrame to a RemoveFrame, or vise versa).
```
Comment (bridiver) [inline chromium_src/third_party/blink/renderer/core/dom/document.cc:36]: Remove -> Remote

File: chromium_src/third_party/blink/renderer/core/dom/document.cc [hunk 6]
```diff
@@ -18,10 +18,42 @@
...
+std::string TopETLDPlusOneForDoc(const Document& doc) {
+  const auto* top_loc_frame = doc.GetFrameOfMasterDocument();
+  if (top_loc_frame == nullptr) {
+    LOG(ERROR) << "Unable to farble frame bc can't find top local frame: "
```
Comment (bridiver) [inline chromium_src/third_party/blink/renderer/core/dom/document.cc:40]: should we have a DCHECK here to see if it's still happening with this change or do we still expect it to be null in some cases?

**Pull request merged**